### ISSUE TITLE
Question: Resolving contexts and variables

### ISSUE THREAD

**Issue by dferber90** [eslint/eslint#4094]
*First of all, if this is not the correct place to ask questions, please point me to it. I have looked for hints on where to ask questions but not found any.*

I'm currently writing an ESLint plugin for Meteor. I'm looking for advice on the best approach to the following two problems regarding rules.

Meteor exposes API through the context of a callback function. In the following example `this.userId` is part of the API exposed by Meteor. There are others like `this.added(..)`, `this.changed(..)`, and so on.

```js
Meteor.publish('foo', function () {
  return Bar.find({
    _id: this.userId // this is the Meteor API I want to lint
  })
})
```

**1: How do I check whether `this` refers to the publication function?**
My approach would be to mark publication functions with `node.isMeteorPublish = true` in every rule `CallExpression`. Then I'd verify `context.getScope().block.isMeteorPublish` to be true in the `MemberExpression`. Is it bad practice to define custom properties on nodes? Should I store this information within the rule instead, like in an array?

**2: How do I resolve variables**
The following is a common pattern in Meteor, which I simplified for this example.

```js
Meteor.publish('foo', function () {
  var self = this
  return Bar.find({
    _id: self.userId // this is the Meteor API I want to lint
  })
})
```

How do I go about resolving `self` to `this`? I have tried using `context.getScope().set.get('self')` and `context.getScope().resolve(identifier)`, but it seems both approaches fail because they give all assignments to `self`, instead of the ones up to the point being linted, so I'm seeing the final value of `self` instead of the current value.

I'm looking for pointers to documentation, API methods I should look at or examples by other rules.
Thank you for ESLint and for taking the time to answer this! :)

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
There's an eslint mailing list that is probably more appropriate for questions.

**Comment by ilyavolodin** [comment]
@dferber90 as @michaelficarra pointed out, mailing list is probably a better place for this kind of questions. There are a long and a short answer to your questions. Short answer is - you can't. ESLint is a static analysis tool, so no execution of linted code occurs and hance you can't track dynamic values of the variables. There are ways you can try to fool the system, but they are all going to be error-prone. If you would like to discuss possible approaches, please send an email to our [mailing list](https://groups.google.com/group/eslint)

**Comment by dferber90** [comment]
I'm aware of the limitations of static analysis and happy to settle with a *close enough* approach.
Thank you for your responses :+1: I'll consider the mailing list if I can't work it out.
